### ISSUE TITLE
Searching the big 3 - make that doable?

### ISSUE THREAD

**Issue by nomadscientist** [galaxyproject/galaxy#16846]
This widget Is amazing: https://singlecell.usegalaxy.eu
<img width="764" alt="Screenshot 2023-10-12 at 23 52 55" src="https://github.com/galaxyproject/galaxy/assets/44605769/adb8d38f-afa3-40d3-bed9-c298aa23fe5e">

This looks across instances! Amazing! Can we please make that appear in the actual Workflows search area? Otherwise I have to direct students/users to the widget rather than the giant WORKFLOWS button in Galaxy...

**Comment by hexylena** [comment]
(links will work after https://github.com/galaxyproject/training-material/pull/4421) but yea it was pretty easy. Would be ace to implement cross-.* search. I'd imagine the following:

- Administrators choose which instances they want to search in the config
- When a user issues a search it also sends the query to those instances, **with a timeout**
- Once the results are collected, display them for the user. (maybe two endpoints, a "local workflows" and "remote workflows" query endpoint), and then separated into two similarly named sections
- Import button should import directly from the remote server
- Could be a nice opportunity to unify the WorkflowHub + Dockstore + Local workflow search into one big workflow search experience.

**Comment by mvdbeek** [comment]
I see how that's cool, but it also gives me big headaches concerning the question: which version is the latest, and which one is the source of truth. I feel like that's answerable if there is something on dockstore/workflowhub, but that's completely unmanagable (at this point, barring any clever ideas) with search results from other Galax instances.

**Comment by mvdbeek** [comment]
(plus the different toolset question, but that's a problem with dockstore/workflow hub as well and something that seems much easier to solve)

**Comment by nomadscientist** [comment]
If there was a column of 'imports' or 'downloads' or 'verified on X' [with a link on how a user could get THEIR workflow verified], or any type of quality-control mechanism on a column, would that solve this issue?

**Comment by mvdbeek** [comment]
Why not make that verification an upload to workflowhub or dockstore ?

**Comment by hexylena** [comment]
> I see how that's cool, but it also gives me big headaches concerning the question: which version is the latest, and which one is the source of truth.

I guess "source of truth" is less of a concern for me? If i'm looking for e.g. a covid-19 workflow, and I get back results that are

- workflows on this server (which have a strong implication of the ability to run immediately)
- and if I scroll down, workflows in other places (WFH/DS/Other servers)

I reckon most of the times I'll take the workflow that's on this server, or I'll be looking for some quality metric like the author's name, if I see IWC it'll be the same as seeing IUC. (yes someone could name squat iwc on every .* server, which was always the case, so, doesn't realy make it worse?). And even then since the *cost* of executing a workflow is low to me (it doesn't expose me to risk), if I try a couple from different users until I find one that works, that's totally fine. it's still an improvement

If you wanted to mark everything on WFH/DS as 'verified' that's fine, I think, as long as we're still getting results from other servers that might be useful to users.

As for 'latest' version, I guess that's really only applies if you've got the same workflow on multiple servers, in different states of up-to-date-ness? Is that something that's happening a lot in practice and wouldn't be solved by the 'verified' badge on WFH/DS?

**Comment by mvdbeek** [comment]
> * workflows on this server (which have a strong implication of the ability to run immediately)
> * and if I scroll down, workflows in other places (WFH/DS/Other servers)

I think it should be WFH/Dockstore first, since someone thinks they're good enough to be published, vs someone just shared a workflow for debugging.

> And even then since the _cost_ of executing a workflow is low to me

You may think Galaxy is not good though when you get workflows that are not good. Plus it'll take a lot of time getting input data right and waiting time for your workflow to run only to realize it's not a good one. That to me is a high cost.

 > As for 'latest' version, I guess that's really only applies if you've got the same workflow on multiple servers, in different states of up-to-date-ness? Is that something that's happening a lot in practice

All the time for collaborative projects like VGP, but even the IWC workflows that you can find on each instance ... they should ideally always come from the source so we can tell people "there's a new version available, do you want to use this one instead ?" which I don't think we can ever do with workflows from other servers.

**Comment by hexylena** [comment]
> I think it should be WFH/Dockstore first,

ok can agree to that.

> You may think Galaxy is not good though when you get workflows that are not good.

but you already have that experience, this doesn't move the bar for that at all. 

if anything this provides a better possible experience because WFH is included, and, this at least gives you more rubbish to look through, rather than being limited if you're on a server that doesn't have a very active community.

> All the time for collaborative projects like VGP, but even the IWC workflows that you can find on each instance 

but those are all under individual accounts, who are working on them, right? not a vgp account on each server? so youd have the a) workflowhub indication and b) from vgp indication

I think any implementation that requires a linear history of workflows worked on across servers is way out of scope and not a goal? we just want more workflows to look through.

> which I don't think we can ever do with workflows from other servers.

no agreed there.

**Comment by mvdbeek** [comment]
> no agreed there.

that needs explanation then, I can't see how that would work, but if you got a plan please let us know

**Comment by hexylena** [comment]
I mean, it wouldn't right. I don't think, for my use cases, that I need a popup of "there's a new version somewhere else from this workflow you imported and modified from someone".

 If I want to know that I'll go re check and reimport, and re-make any changes I made to it.

**Comment by mvdbeek** [comment]
Yeah, but if people import the non-source version we lose all links to the upstream source. I have no problem with the search as long as it doesn't actively prevent linked workflows.

**Comment by nomadscientist** [comment]
Huh? Linked workflows?

**Comment by nomadscientist** [comment]
From what I could follow, I agree with the 'we just want more workflows to sort through'. It's easy to re-order by 'date modified/shared' which takes care of dealing with a lot of the rubbish/older versions of stuff

**Comment by mvdbeek** [comment]
> Linked workflows?

You have a source you're working on, this is what everything else should be based on and what your collaborators should use. Now you try this at a different server. The workflow will be newer, you try a few things but in the end you don't care about these changes. That workflow will now be newer with no indication to anyone else that you should probably not use this.

In return, if there is an associated TRS id and TRS version, we can track the lineage and know the source and that maybe there has been an update you want to pull in to get the latest and greatest version.
